debian/nctoh5.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH NCTOH5 1 "November 11, 2004"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
nctoh5 \- Convert a NetCDF file into a PyTables (HDF5 format) file
.SH SYNOPSIS
.B nctoh5
.RB \-h
.RB \-v
.RB \-o
.RB \-\-complevel=(0\-9)
.RB \-\-complib=lib
.RB \-\-shuffle=(0|1)
.RB \-\-fletcher32=(0|1)
.RB \-\-unpackshort=(0|1)
.RB \-\-quantize=(0|1)
.RB netcdffilename
.RB hdf5filename
.br
.SH DESCRIPTION
Convert a generic NetCDF file into a PyTables (HDF5 format) file
.SH OPTIONS
A summary of options is included below.
.TP
.B \-h
Prints a help text.
.TP
.B \-v
Show more information.
.TP
.B \-o
Overwite destination file.
.TP
.BI \-\-complevel=(0-9)
Sets a compression level (0 for no compression, which is the default).
.TP
.BI \-\-complib=lib
Sets the compression library to be used during the copy. \fIlib\fR can
be set to "zlib", "lzo" or "ucl". Defaults to "zlib".
.TP
.BI \-\-shuffle=(0|1)
Activate or not the shuffling filter (default is active if complevel>0).
.TP
.BI \-\-fletcher32=(0|1)
Whether activate or not the fletcher32 filter (not active by default).
.TP
.BI \-\-unpackshort=(0|1)
Unpack short integer variables to float variables using scale_factor
and add_offset netCDF variable attributes (not active by default).
.TP
.BI \-\-quantize=(0|1)
Quantize data to improve compression using least_significant_digit
netCDF variable attribute (not active by default).  See
.br
.nh
http://www.cdc.noaa.gov/cdc/conventions/cdc_netcdf_standard.shtml
.br
.hy
for further explanation of what this attribute means.

.br

.SH SEE ALSO
.BR ptdump (1), ptrepack (1).
.br
These utilities are documented fully by
.IR "PyTables user's manual".
.SH AUTHOR
This manual page was written by Francesc Alted <faltet@pytables.com>.


debian/changelog
pytables (2.0pro-1) unstable; urgency=low

  * New upstream release.

 -- Francesc Altet <faltet@carabos.com>  Thu, 05 Jul 2007 17:18:35 +0200



tables/netcdf3/scripts/nctoh5.py
"""
convert netCDF file to HDF5 using Scientific.IO.NetCDF and PyTables.
Jeff Whitaker <jeffrey.s.whitaker@noaa.gov>

Added some flags to select filters, as well as some small improvements.
Francesc Alted <faltet@carabos.com>

This requires Scientific from
http://starship.python.net/~hinsen/ScientificPython
"""

import sys, os.path, getopt, time

import tables.netcdf3

from tables.leaf import Filters


def nctoh5(ncfilename, h5filename, filters, verbose, overwritefile):
    # open h5 file
    if overwritefile:
        h5file = tables.netcdf3.NetCDFFile(h5filename, mode = "w")
    else:
        h5file = tables.netcdf3.NetCDFFile(h5filename, mode = "a")
    # convert to netCDF
    nobjects, nbytes = h5file.nctoh5(ncfilename,filters=filters)
    # ncdump-like output
    if verbose:
        print 'contents of hdf5 file:'
        print '----------------------'
        print h5file
    # Close the file
    h5file.close()
    return nobjects, nbytes


def main():
    if not tables.netcdf3.ScientificIONetCDF_imported:
        sys.stderr.write(
            'You need Scientific Python installed in order to use this utility.\n')
        sys.exit(1)

    usage = """usage: %s [-h] [-v] [-o] [--complevel=(0-9)] [--complib=lib] [--shuffle=(0|1)] [--fletcher32=(0|1)] netcdffilename hdf5filename
     -h -- Print usage message.
     -v -- Show more information.
     -o -- Overwite destination file.
     --complevel=(0-9) -- Set a compression level (0 for no compression, which
         is the default).
     --complib=lib -- Set the compression library to be used during the copy.
         lib can be set to "zlib", "lzo" or "ucl". Defaults to "zlib".
     --shuffle=(0|1) -- Activate or not the shuffling filter (default is active
         if complevel>0).
     --fletcher32=(0|1) -- Whether to activate or not the fletcher32 filter (not
         active by default).
    \n""" % os.path.basename(sys.argv[0])

    try:
        opts, pargs = getopt.getopt(sys.argv[1:], 'hvo',
                                    ['complevel=',
                                     'complib=',
                                     'shuffle=',
                                     'fletcher32=',
                                     ])
    except:
        (type, value, traceback) = sys.exc_info()
        print "Error parsing the options. The error was:", value
        sys.stderr.write(usage)
        sys.exit(0)

    # default options
    verbose = 0
    overwritefile = 0
    complevel = None
    complib = None
    shuffle = 0
    fletcher32 = 0

    # Get the options
    for option in opts:
        if option[0] == '-h':
            sys.stderr.write(usage)
            sys.exit(0)
        elif option[0] == '-v':
            verbose = 1
        elif option[0] == '-o':
            overwritefile = 1
        elif option[0] == '--complevel':
            complevel = int(option[1])
        elif option[0] == '--complib':
            complib = option[1]
        elif option[0] == '--shuffle':
            shuffle = int(option[1])
        elif option[0] == '--fletcher32':
            fletcher32 = int(option[1])
        else:
            print option[0], ": Unrecognized option"
            sys.stderr.write(usage)
            sys.exit(0)

    # if we pass a number of files different from 2, abort
    if len(pargs) <> 2:
        print "You need to pass both source and destination!."
        sys.stderr.write(usage)
        sys.exit(0)

    # Catch the files passed as the last arguments
    ncfilename = pargs[0]
    h5filename = pargs[1]


    # Build the Filters instance
    if complevel==None and complib==None and shuffle==0 and fletcher32==0:
        filters = None
    else:
        if complevel is None: complevel = 0
        if complib is None: complib = "zlib"
        if fletcher32 is None: fletcher32 = 0
        filters = Filters(complevel=complevel, complib=complib,
                          shuffle=shuffle, fletcher32=fletcher32)

    # Some timing
    t1 = time.time()
    cpu1 = time.clock()
    # Copy the file
    if verbose:
        print "+=+"*20
        print "Starting conversion from %s to %s" % (ncfilename, h5filename)
        if filters == None:
            print "Using default filters (complevel=6,complib='zlib',shuffle=1,fletcher32=0)"
        else:
            print "Applying filters:", filters
        print "+=+"*20

    # Do the conversion
    (nobjects, nbytes) = nctoh5(ncfilename, h5filename, filters, verbose, overwritefile)

    # Gather some statistics
    t2 = time.time()
    cpu2 = time.clock()
    tcopy = round(t2-t1, 3)
    cpucopy = round(cpu2-cpu1, 3)
    tpercent = int(round(cpucopy/tcopy, 2)*100)
    if verbose:
        print "Number of variables copied:", nobjects
        print "KBytes copied:", round(nbytes/1024.,3)
        print "Time copying: %s s (real) %s s (cpu)  %s%%" % \
              (tcopy, cpucopy, tpercent)
        print "Copied variable/sec: ", round(nobjects / float(tcopy),1)
        print "Copied KB/s :", int(nbytes / (tcopy * 1024))
